pkg/pillar/cmd/logmanager/logmanager.go

```diff
@@ -536,81 +532,99 @@ func processEvents(image string, prevLastSent time.Time,
 
 			log.Debugf("processEvents(%s): sending at messageCount %d, byteCount %d\n",
 				image, messageCount, byteCount)
-			sent = sendProtoStrForLogs(reportLogs, image,
-				iteration, eveVersion)
-			messageCount = 0
-			iteration++
-			if sent {
-				recordLast(lastSentDirname, image)
+			if event.isAppLog {
+				sent = sendProtoStrForAppLogs(appUUID, appLogBundle, iteration, image)
+				delete(appLogBundles, appUUID)
 			} else {
-				recordLast(lastDeferDirname, image)
+				sent = sendProtoStrForLogs(reportLogs, image,
+					iteration, eveVersion)
+			}
+			iteration++
+			if !sent {
 				deferInprogress = true
 			}
 
 		case <-flushTimer.C:
-			if messageCount == 0 {
-				break
-			}
 			log.Debugf("processEvents(%s) flush at %s dropped %d messageCount %d bytecount %d\n",
 				image, time.Now().String(),
 				dropped, messageCount,
 				proto.Size(reportLogs))
-			sent := sendProtoStrForLogs(reportLogs, image,
-				iteration, eveVersion)
-			messageCount = 0
+			// Iterate the app log bundle map and send out all app logs
+			sent := flushAllLogBundles(image, iteration, eveVersion, reportLogs, appLogBundles)
 			iteration++
-			if sent {
-				recordLast(lastSentDirname, image)
-			} else {
-				recordLast(lastDeferDirname, image)
+			if !sent {
 				deferInprogress = true
 			}
 		}
 	}
 }
 
-// Touch/create a file to keep track of when things where sent before a reboot
-func recordLast(dirname string, image string) {
-	log.Debugf("recordLast(%s, %s)\n", dirname, image)
-	filename := fmt.Sprintf("%s/%s/%s", types.PersistDir, dirname, image)
-	_, err := os.Stat(filename)
-	if err != nil {
-		file, err := os.Create(filename)
-		if err != nil {
-			log.Infof("recordLast: %s\n", err)
-			return
+func flushAllLogBundles(image string, iteration int, eveVersion string,
+	reportLogs *logs.LogBundle, appLogBundles map[string]*logs.AppInstanceLogBundle) bool {
+	sent := sendProtoStrForLogs(reportLogs, image, iteration, eveVersion)
+	if !sent {
+		return false
+	}
+	appBundlesToDelete := []string{}
+	for appUUID, appLogBundle := range appLogBundles {
+		log.Debugf("flushAllLogBundles: Trying to flush App bundle with UUID %s and %d logs",
+			appUUID, len(appLogBundle.Log))
+		sent = sendProtoStrForAppLogs(appUUID, appLogBundle, iteration, image)
+		if !sent {
+			break
 		}
-		file.Close()
+		log.Debugf("flushAllLogBundles: Flushed App bundle with UUID %s", appUUID)
+		appBundlesToDelete = append(appBundlesToDelete, appUUID)
 	}
-	_, err = os.Stat(filename)
-	if err != nil {
-		log.Errorf("recordLast: %s\n", err)
-		return
+	for _, appUUID := range appBundlesToDelete {
+		delete(appLogBundles, appUUID)
 	}
-	now := time.Now()
-	err = os.Chtimes(filename, now, now)
-	if err != nil {
-		log.Errorf("recordLast: %s\n", err)
+	return sent
+}
+
+var msgIDCounter = 1
```

Comment:
Don't you want this per app? Don't know how it gets used on the server side but it would potentially be confusing to see gaps just because there are multiple app instances.

Reply:
I've brought this up in meetings earlier about the relevance of this field while ordering messages. Turns out nobody seems to care and know how this can be used. BTW if the software tool itself can order ascending/descending fashion, gaps should not make any difference. If it's not used by consumer (cloud), it makes no sense to put any more complexity in device IMHO.

Reply:
@nordmark you are right. Please forgive my ignorance. Yes, even if they are being channeled to different endpoints, cloud will see gaps in message ids. But, if cloud can maintain the order, we should be good,

Reply:
Including it with gaps is just confusiong; either it shouldn't be there or it should have some useful meaning
See above for suggestion